assignment5.cpp

```diff
@@ -0,0 +1,219 @@
+        
+        if (first_not_common_chars.first != not_found) {
+            char first = first_not_common_chars.first;
+            char second = first_not_common_chars.second;
+            
+            if (!adjacency_matrix.count(first)) {
+                vector<char> adjacent_chars;
+                adjacent_chars.push_back(second);
+                adjacency_matrix.insert(make_pair(first, adjacent_chars));
+            } else {
+                adjacency_matrix[first].push_back(second);
+            }
+            
+        }
+        
+    }
+    
+    return adjacency_matrix;
+    
+}
+
+/* finds the in degrees of the nodes in the dag */
+map<char, int> find_in_degrees(map<char, vector<char> > adjacency_matrix) {
+    map<char, int> in_degrees;
+    
+    for (map<char, vector<char> >::iterator it = adjacency_matrix.begin(); it!=adjacency_matrix.end(); ++it) {
+        vector<char> adj_chars = it -> second;
+        for (int i = 0; i < adj_chars.size(); i++) {
+            if (!in_degrees.count(adj_chars[i])) {
+                in_degrees.insert(make_pair(adj_chars[i], 1));
+            } else {
+                in_degrees[adj_chars[i]]++;
+            }
+        }
+    }
+    
+    return in_degrees;
+}
+
+/* visits node in the dag */
+void visit(char char_to_visit, map<char, int> *in_degrees, set<char> *visited, map<char, vector<char> > adj_matrix, queue<char> *roots_of_dag) {
+    vector<char> adj_chars = adj_matrix[char_to_visit];
+    
+    visited->insert(char_to_visit);
+    
+    for (int i = 0; i < adj_chars.size(); i++) {
+        if (!visited->count(adj_chars[i])) {
+            (*in_degrees)[adj_chars[i]]--;
+            if ((*in_degrees)[adj_chars[i]] == 0) {
+                roots_of_dag->push(adj_chars[i]);
+            }
+        }
+    }
+}
+
+/* does the topological sorting and returns the ordered alphabet */
+vector<char> topological_sort(map<char, vector<char> > adjacency_matrix, map<char, int> in_degrees, vector<string> dictionary) {
+    vector<char> ordered_alphabet;
+    set<char> visited;
+    queue<char> roots_of_dag;
+    
+    /* adds all chars with 0 in degree to the queue */
+    for (int i = 0; i < dictionary.size(); i++) {
+        for (int j = 0; j < dictionary[i].size(); j++) {
+            if (!in_degrees.count(dictionary[i][j])) {
+                roots_of_dag.push(dictionary[i][j]);
+            }
+        }
+    }
+    
+    while (!roots_of_dag.empty()) {
+        char curr_node = roots_of_dag.front();
+        roots_of_dag.pop();
+        if (!visited.count(curr_node)) {
+            ordered_alphabet.push_back(curr_node);
+            visit(curr_node, &in_degrees, &visited, adjacency_matrix, &roots_of_dag);
+        }
+    }
+    
+
+    
+    return ordered_alphabet;
+}
+
+/* returns ordered alphabet given a dictionary */
+vector<char> alphabet(vector<string> dictionary) {
+    vector<char> ordered_alphabet;
+    map<char, vector<char> > adjacency_matrix;
```

Comment:
Define local variables when you use them, not before.
`map<char, vector<char> > adjacency_matrix = build_adjacency_matrix(dictionary);`
and
`vector<char> ordered_alphabet = topological_sort(adjacency_matrix, in_degrees, dictionary);`

Instead of defining at the beginning of the function.